xds/src/test/java/io/grpc/xds/CsdsServiceTest.java

```diff
@@ -0,0 +1,891 @@
+import io.envoyproxy.envoy.admin.v3.RoutesConfigDump.DynamicRouteConfig;
+import io.envoyproxy.envoy.admin.v3.UpdateFailureState;
+import io.envoyproxy.envoy.config.cluster.v3.Cluster;
+import io.envoyproxy.envoy.config.core.v3.Node;
+import io.envoyproxy.envoy.config.endpoint.v3.ClusterLoadAssignment;
+import io.envoyproxy.envoy.config.listener.v3.Listener;
+import io.envoyproxy.envoy.config.route.v3.RouteConfiguration;
+import io.envoyproxy.envoy.service.status.v3.ClientConfig;
+import io.envoyproxy.envoy.service.status.v3.ClientStatusDiscoveryServiceGrpc;
+import io.envoyproxy.envoy.service.status.v3.ClientStatusRequest;
+import io.envoyproxy.envoy.service.status.v3.ClientStatusResponse;
+import io.envoyproxy.envoy.service.status.v3.PerXdsConfig;
+import io.envoyproxy.envoy.type.matcher.v3.NodeMatcher;
+import io.grpc.Status;
+import io.grpc.Status.Code;
+import io.grpc.StatusRuntimeException;
+import io.grpc.internal.ObjectPool;
+import io.grpc.internal.testing.StreamRecorder;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.GrpcServerRule;
+import io.grpc.xds.AbstractXdsClient.ResourceType;
+import io.grpc.xds.XdsClient.ResourceMetadata;
+import io.grpc.xds.XdsClient.ResourceMetadata.ResourceMetadataStatus;
+import io.grpc.xds.XdsNameResolverProvider.XdsClientPoolFactory;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.runners.Enclosed;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link CsdsService}. */
+@RunWith(Enclosed.class)
+public class CsdsServiceTest {
+  private static final String NODE_ID =
+      "projects/42/networks/default/nodes/5c85b298-6f5b-4722-b74a-f7d1f0ccf5ad";
+  private static final EnvoyProtoData.Node BOOTSTRAP_NODE =
+      EnvoyProtoData.Node.newBuilder().setId(NODE_ID).build();
+  private static final XdsClient XDS_CLIENT_NO_RESOURCES = new XdsClient() {
+    @Override
+    EnvoyProtoData.Node getNode() {
+      return BOOTSTRAP_NODE;
+    }
+
+    @Override
+    String getCurrentVersion(ResourceType type) {
+      return "getCurrentVersion." + type.name();
+    }
+
+    @Override
+    Map<String, ResourceMetadata> getSubscribedResourcesMetadata(ResourceType type) {
+      return ImmutableMap.of();
+    }
+  };
+
+  @RunWith(JUnit4.class)
+  public static class ServiceTests {
+    private static final CsdsService CSDS_SERVICE_MINIMAL =
+        new CsdsService(new FakeXdsClientPoolFactory(XDS_CLIENT_NO_RESOURCES));
+    private static final ClientStatusRequest REQUEST = ClientStatusRequest.getDefaultInstance();
+    private static final ClientStatusRequest REQUEST_INVALID =
+        ClientStatusRequest.newBuilder().addNodeMatchers(NodeMatcher.getDefaultInstance()).build();
+
+    @Rule public final GrpcServerRule grpcServerRule = new GrpcServerRule().directExecutor();
```

Comment:
It's generally [favorable](https://github.com/grpc/grpc-java/blob/ad2b26436ccd64cb394127ebc0cd3323817dd8a3/testing/src/main/java/io/grpc/testing/GrpcServerRule.java#L38-L44) to use [`GrpcCleanupRule`](https://github.com/grpc/grpc-java/blob/master/testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java). It's a newer thing that does everything GrpcServerRule can do plus a lot [extra](https://github.com/grpc/grpc-java/pull/4394#issue-184194598).

I am still fine with using GrpcServerRule.

Reply:
I think, at least for as tests work now, `GrpcServerRule` is a really good fit. `CsdsServiceTest.ServiceTests` follow the same pattern - create server with channel to it, add service to the registry, make a request, test it. Changing `GrpcCleanupRule` would pretty much replicate this exact behavior provided by `GrpcServerRule` out of box.